Improve ConstantExpr performance

kren1 opened the pull request:
I did some flame graph profiling and noticed that ConstantExprs take some time, because they are computing the hash.

The reason is that `hash_value(APInt)` is quite expensive. Considering that most of our ConstantExpr, should fit in 64bits, it should be fine to just use `getLimitedValue`. If there are some ConstantExpr with more than 64bits, this still works fine, but there might be more collisions. However I don't think wide ConstantExprs are  common enough to be a problem.

kren1 commented:
The `test/Feature/DeterministicSwitch.c` test failed, because the order of output changed. I think this makes sense if hash values change and the test name seems to imply it only requires determinism, so I think it's OK to change.

ccadar commented:
@kren1 , I'm happy with this change.  Can you please squash the two commits with a message along the lines: "Change ConstantExpr hashing function to a faster version and modified affected test"?

kren1 commented:
@ccadar Done

kren1 commented:
@MartinNowack but we scalarize vector instructions, so that doesn't count.

Sure you concatenate expression, but I can't think of a case where they wouldn't fit in 64bits. It's my understanding that Exprs need to fit into registers, so they can't be larger than 64bit? 

The question is how common are wider than 64bit (constant) expressions. I haven't seen one yet. Also note that this is a better hash function for less thab 64bit exprs, so it might intact reduce collisions.

I can show you my graphs on Monday.

kren1 commented:
@MartinNowack I think I addressed your concerns by only using the fast hash function if width is less than 64 bits, otherwise it's the old behaviour. It still looks faster on my benchmark.

ccadar commented on lib/Expr/Expr.cpp:
```diff
@@ -182,7 +182,12 @@ unsigned Expr::computeHash() {
 unsigned ConstantExpr::computeHash() {
-  hashValue = hash_value(value) ^ (getWidth() * MAGIC_HASH_CONSTANT);
+  Expr::Width w = getWidth();
+  if(w <= 64)
```
Can you add a space before the open parenthesis?

ccadar commented:
Thanks, @kren1 , I'm happy with this. 
@MartinNowack ?

MartinNowack commented:
@ccadar I'm happy with this as well.

Just to add a few comments to this PR - in case somebody else reads it.
The width of an expression tree can be arbitrary - including constant expressions. Of course, it's more likely to have small-width expression than those with large width.
To cope with size limitations, like registers, an extract expression is used to adapt the arbitrary wide expression to a specific small size.

MartinNowack commented:
@kren1 Thanks for the PR - going to merge ;)
